counter() integration with counter-styles

vitorroriz opened the pull request:
#### b1867c1e91684237ff93db475bf762fdf4b08536
<pre>
counter() integration with counter-styles
<a href="https://bugs.webkit.org/show_bug.cgi?id=253920">https://bugs.webkit.org/show_bug.cgi?id=253920</a>
rdar://106084053

Integrating counter() function with counter-styles defined.
We are also updating the counter() behavior regarding &apos;none&apos;.
Before, &apos;none&apos; would generate a empty string counter. However,
the spec was updated and &apos;none&apos; no longer generates a valid counter.

Reviewed by Tim Nguyen.

* LayoutTests/TestExpectations:
* LayoutTests/css2.1/t1202-counter-16-f.html:
* LayoutTests/css2.1/t1202-counters-18-f.html:
- These tests are updated because list-style-type now accepts a custom-ident.
Therefore, the custom-ident uknowntype no longer produces a invalid counter,
since it will point to a counter-style that does not exist, which falls back to decimal.
that does not exist, we should fallback to decimal.

* LayoutTests/fast/css/counters/counter-cssText-expected.txt:
* LayoutTests/fast/css/counters/counter-cssText.html:
* LayoutTests/fast/css/counters/getCounterValue-expected.txt:
* LayoutTests/fast/css/counters/getCounterValue.html:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-cascade/layer-counter-style-override-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/layer-counter-style-override-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/content-invalid-expected.txt:
- Updating expectations for &apos;none&apos; not generating a valid counter.

* LayoutTests/platform/glib/css2.1/t1202-counter-01-b-expected.txt: Removed.
* LayoutTests/platform/glib/css2.1/t1202-counters-01-b-expected.txt: Removed.
* LayoutTests/css2.1/t1202-counter-01-b.html: Removed.
* LayoutTests/css2.1/t1202-counters-01-b.html: Removed.
* LayoutTests/platform/gtk/css2.1/t1202-counter-01-b-expected.png: Removed.
* LayoutTests/platform/gtk/css2.1/t1202-counters-01-b-expected.png: Removed.
* LayoutTests/platform/ios/css2.1/t1202-counter-01-b-expected.txt: Removed.
* LayoutTests/platform/ios/css2.1/t1202-counters-01-b-expected.txt: Removed.
* LayoutTests/platform/mac/css2.1/t1202-counter-01-b-expected.png: Removed.
* LayoutTests/platform/mac/css2.1/t1202-counter-01-b-expected.txt: Removed.
* LayoutTests/platform/mac/css2.1/t1202-counters-01-b-expected.png: Removed.
* LayoutTests/platform/mac/css2.1/t1202-counters-01-b-expected.txt: Removed.
* LayoutTests/platform/wincairo/css2.1/t1202-counter-01-b-expected.txt: Removed.
* LayoutTests/platform/wincairo/css2.1/t1202-counters-01-b-expected.txt: Removed.
- We are removing 2 tests and its -expected files since these were covering the
outdated counter(none) behavior. The invalid behavior is already covered by
LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/content-invalid.html

* Source/WebCore/css/CSSCounterStyle.cpp:
(WebCore::CSSCounterStyle::text):
* Source/WebCore/css/CSSCounterStyleRegistry.cpp:
(WebCore::CSSCounterStyleRegistry::resolvedCounterStyle):
* Source/WebCore/css/CSSCounterStyleRegistry.h:
* Source/WebCore/css/CSSCounterValue.cpp:
(WebCore::CSSCounterValue::CSSCounterValue):
(WebCore::CSSCounterValue::create):
(WebCore::CSSCounterValue::equals const):
(WebCore::CSSCounterValue::customCSSText const):
(WebCore::CSSCounterValue::counterStyleCSSText const):
* Source/WebCore/css/CSSCounterValue.h:
(WebCore::CSSCounterValue::counterStyle const):
(WebCore::CSSCounterValue::listStyle const): Deleted.
* Source/WebCore/css/DeprecatedCSSOMCounter.h:
* Source/WebCore/css/DeprecatedCSSOMPrimitiveValue.cpp:
(WebCore::DeprecatedCSSOMPrimitiveValue::getCounterValue const):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeCounterContent):
(WebCore::CSSPropertyParserHelpers::consumeContent):
* Source/WebCore/rendering/RenderCounter.cpp:
(WebCore::RenderCounter::originalText const):
(WebCore::RenderCounter::counterStyle const):
* Source/WebCore/rendering/RenderCounter.h:
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::listMarkerText):
(WebCore::RenderListMarker::counterStyle const):
* Source/WebCore/rendering/style/CounterContent.h:
(WebCore::CounterContent::CounterContent):
(WebCore::CounterContent::listStyleType const):
(WebCore::operator==):
(WebCore::CounterContent::listStyle const): Deleted.
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueContent):

Canonical link: <a href="https://commits.webkit.org/261985@main">https://commits.webkit.org/261985@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e5eacc0a061f8dca9a1496d4d4cebfd138dc2e82

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/169 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/179 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/186 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/171 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/159 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/170 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/173 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/169 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/173 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/168 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/187 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/164 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/153 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/148 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/172 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/180 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/161 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/142 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/169 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/38 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/166 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSCounterStyleRegistry.h:
```diff
@@ -41,6 +42,7 @@ class CSSCounterStyleRegistry {
     WTF_MAKE_FAST_ALLOCATED;
 public:
     RefPtr<CSSCounterStyle> resolvedCounterStyle(const AtomString&);
```
I think this is unused?

nt1m commented on Source/WebCore/css/CSSCounterValue.cpp:
```diff
@@ -28,32 +28,33 @@
     , m_identifier(WTFMove(identifier))
     , m_separator(WTFMove(separator))
-    , m_listStyle(listStyle)
+    , m_listStyleValue(WTFMove(listStyleValue))
```
can we name this and friends to be `listStyleType`? because this is effectively what this represents.

nt1m commented on Source/WebCore/css/DeprecatedCSSOMPrimitiveValue.cpp:
```diff
@@ -130,7 +130,7 @@ ExceptionOr<Ref<DeprecatedCSSOMCounter>> DeprecatedCSSOMPrimitiveValue::getCount
         return Exception { InvalidAccessError };
     auto& value = downcast<CSSCounterValue>(m_value.get());
-    return DeprecatedCSSOMCounter::create(value.identifier(), value.separator(), value.listStyle());
+    return DeprecatedCSSOMCounter::create(value.identifier(), value.separator(), value.listStyleValue()->valueID());
```
Is this always correct? I think valueID won't always exist or be valid, if the type is string or custom ident.

vitorroriz replied:
using `isValueID()` is fine because it returns `CSSValueInvalid` when the CSSValue is not a ValueID, however I need to transform the lambda I created for converting to string into a function so I can re-use it here.

nt1m commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -1562,10 +1562,11 @@ inline void BuilderCustom::applyValueContent(BuilderState& builderState, CSSValu
             auto& counter = downcast<CSSCounterValue>(item);
-            auto listStyle = fromCSSValueID<ListStyleType::Type>(counter.listStyle());
-            builderState.style().setContent(makeUnique<CounterContent>(counter.identifier(), listStyle, counter.separator()), didSet);
+            ListStyleType listStyleType { ListStyleType::Type::Decimal, nullAtom() };
```
This should be `{ ListStyleType::Type::CounterStyle, "decimal"_s }` eventually, but it can be changed later on.

vitorroriz replied:
Actually, I've realized this should be set at the parser level to a CSSValueDecimal like before, and not here. I'm changing it. Then in it will correctly be converted to the right ListStyleType

nt1m commented on Source/WebCore/rendering/RenderCounter.cpp:
```diff
@@ -463,13 +464,14 @@ String RenderCounter::originalText() const
 
-    String text = listMarkerText(m_counter.listStyle(), value);
+    auto counterStyle = this->counterStyle();
+    String text = listMarkerText(m_counter.listStyleType().type, value, counterStyle.get());
```
do you need to call `.get()` here on `counterStyle`?

vitorroriz replied:
I wanted listMarkerText() to receive a counterStyle ptr and not a RefPtr because it shouldn't have to care about the counterStyle life-cycle.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -6568,18 +6568,17 @@ static RefPtr<CSSValue> consumeCounterContent(CSSParserTokenRange args, bool cou
     }
 
-    auto listStyle = CSSValueDecimal;
+    RefPtr<CSSValue> listStyleType = CSSPrimitiveValue::create(CSSValueDecimal);
```
Not sure this will be accurate when @counter-style is on.

vitorroriz replied:
When it's on, predefined values are appropriately converted to ListStyleType with CounterStyle as type and an identifier.

nt1m commented on Source/WebCore/css/CSSCounterValue.cpp:
```diff
@@ -28,32 +28,33 @@
-    auto listStyleSeparator = m_listStyle == CSSValueDecimal ? ""_s : ", "_s;
-    auto listStyleLiteral = m_listStyle == CSSValueDecimal ? ""_s : nameLiteral(m_listStyle);
+    auto listStyleSeparator = m_listStyleType->valueID() == CSSValueDecimal ? ""_s : ", "_s;
+    auto listStyleLiteral = m_listStyleType->valueID() == CSSValueDecimal ? ""_s : nameLiteral(m_listStyleType->valueID());
```
Is valueID() correct for counter-styles that are not in the CSSValueKeywords list?

nt1m replied:
Is `CSSValueInvalid` desired? If we have a non-UA counter style name, it should be reflected.

nt1m commented on Source/WebCore/css/DeprecatedCSSOMPrimitiveValue.cpp:
```diff
@@ -127,10 +127,21 @@ ExceptionOr<String> DeprecatedCSSOMPrimitiveValue::getStringValue() const
+
+        auto& primitiveValue = downcast<CSSPrimitiveValue>(value);
+        ASSERT(primitiveValue.isString());
+        return primitiveValue.stringValue();
```
The counter() function doesn't take string values afaik

nt1m commented on Source/WebCore/rendering/RenderCounter.cpp:
```diff
@@ -463,13 +464,14 @@ String RenderCounter::originalText() const
         if (!child->actsAsReset())
             child = child->parent();
         while (CounterNode* parent = child->parent()) {
-            text = listMarkerText(m_counter.listStyle(), child->countInParent())
```
small nit that can be addressed later, but I think we should make the third argument of `listMarkerText` non-optional (it can be nullptr if needed though).

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -6568,18 +6568,17 @@ static RefPtr<CSSValue> consumeCounterContent(CSSParserTokenRange args, bool cou
     if (consumeCommaIncludingWhitespace(args)) {
-        listStyle = args.peek().id();
-        if (listStyle != CSSValueNone && !isPredefinedCounterStyle(listStyle))
+        listStyleType = consumeListStyleType(args, context);
```
We should exclude CSSValueNone and `<string>` values per the syntax.

nt1m replied:
There should be a test for this btw

vitorroriz replied:
hmmmm interesting, CSSValueNone was being accepted before.

nt1m replied:
Can we add a test for this, preferably in WPT?

nt1m commented on LayoutTests/css2.1/t1202-counter-16-f.html:
```diff
@@ -10,15 +10,14 @@
-    content: counter(f, decimal, ".");
-    content: counter(f, decimal, decimal);
-    content: counter(f, unknowntype);
+    content: counter(success, uknowntype);
```
```suggestion
    content: counter(success, custom-ident);
    content: counter(failure);
```

nt1m commented on LayoutTests/css2.1/t1202-counters-18-f.html:
```diff
@@ -10,16 +10,15 @@
-    content: counters(f, ".", decimal, decimal);
-    content: counters(f, ".", decimal, ".");
-    content: counters(f, ".", unknowntype);
+    content: counters(success, ".", uknowntype);
```
```suggestion
    content: counters(success, ".", custom-ident);
    content: counters(failure, ".");
```

nt1m commented on Source/WebCore/css/CSSCounterValue.cpp:
```diff
@@ -28,32 +28,33 @@
     , m_identifier(WTFMove(identifier))
     , m_separator(WTFMove(separator))
-    , m_listStyle(listStyle)
+    , m_listStyleType(WTFMove(listStyleType))
```
I hate to make you change the name again, but I realize the name should probably be `m_counterStyle` since it only points to a `counter-style` name or `symbols()` (which at the end of the day also represents counter style). `listStyleType` is inaccurate, because we don't take `<string>`, or `none`.

nt1m replied:
(ditto everywhere else)

nt1m commented on Source/WebCore/css/CSSCounterValue.cpp:
```diff
@@ -63,4 +64,17 @@ String CSSCounterValue::customCSSText() const
     return result.toString();
 }
 
+String CSSCounterValue::listStyleTypeAsString() const
```
`counterStyleCSSText` ?

nt1m commented on Source/WebCore/css/CSSCounterValue.cpp:
```diff
@@ -28,32 +28,33 @@
+Ref<CSSCounterValue> CSSCounterValue::create(AtomString identifier, AtomString separator, RefPtr<CSSValue> listStyleType)
 {
-    return adoptRef(*new CSSCounterValue(WTFMove(identifier), WTFMove(separator), listStyle));
+    return adoptRef(*new CSSCounterValue(WTFMove(identifier), WTFMove(separator), WTFMove(listStyleType)));
```
We should also name the parameters `counterStyle`

vitorroriz committed:
counter() integration with counter-styles
https://bugs.webkit.org/show_bug.cgi?id=253920
rdar://106084053

Integrating counter() function with counter-styles defined.
We are also updating the counter() behavior regarding 'none'.
Before, 'none' would generate a empty string counter. However,
the spec was updated and 'none' no longer generates a valid counter.

Reviewed by Tim Nguyen.

* LayoutTests/TestExpectations:
* LayoutTests/css2.1/t1202-counter-16-f.html:
* LayoutTests/css2.1/t1202-counters-18-f.html:
- These tests are updated because list-style-type now accepts a custom-ident.
Therefore, the custom-ident uknowntype no longer produces a invalid counter,
since it will point to a counter-style that does not exist, which falls back to decimal.
that does not exist, we should fallback to decimal.

* LayoutTests/fast/css/counters/counter-cssText-expected.txt:
* LayoutTests/fast/css/counters/counter-cssText.html:
* LayoutTests/fast/css/counters/getCounterValue-expected.txt:
* LayoutTests/fast/css/counters/getCounterValue.html:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-cascade/layer-counter-style-override-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/layer-counter-style-override-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/content-invalid-expected.txt:
- Updating expectations for 'none' not generating a valid counter.

* LayoutTests/platform/glib/css2.1/t1202-counter-01-b-expected.txt: Removed.
* LayoutTests/platform/glib/css2.1/t1202-counters-01-b-expected.txt: Removed.
* LayoutTests/css2.1/t1202-counter-01-b.html: Removed.
* LayoutTests/css2.1/t1202-counters-01-b.html: Removed.
* LayoutTests/platform/gtk/css2.1/t1202-counter-01-b-expected.png: Removed.
* LayoutTests/platform/gtk/css2.1/t1202-counters-01-b-expected.png: Removed.
* LayoutTests/platform/ios/css2.1/t1202-counter-01-b-expected.txt: Removed.
* LayoutTests/platform/ios/css2.1/t1202-counters-01-b-expected.txt: Removed.
* LayoutTests/platform/mac/css2.1/t1202-counter-01-b-expected.png: Removed.
* LayoutTests/platform/mac/css2.1/t1202-counter-01-b-expected.txt: Removed.
* LayoutTests/platform/mac/css2.1/t1202-counters-01-b-expected.png: Removed.
* LayoutTests/platform/mac/css2.1/t1202-counters-01-b-expected.txt: Removed.
* LayoutTests/platform/wincairo/css2.1/t1202-counter-01-b-expected.txt: Removed.
* LayoutTests/platform/wincairo/css2.1/t1202-counters-01-b-expected.txt: Removed.
- We are removing 2 tests and its -expected files since these were covering the
outdated counter(none) behavior. The invalid behavior is already covered by
LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/content-invalid.html

* Source/WebCore/css/CSSCounterStyle.cpp:
(WebCore::CSSCounterStyle::text):
* Source/WebCore/css/CSSCounterStyleRegistry.cpp:
(WebCore::CSSCounterStyleRegistry::resolvedCounterStyle):
* Source/WebCore/css/CSSCounterStyleRegistry.h:
* Source/WebCore/css/CSSCounterValue.cpp:
(WebCore::CSSCounterValue::CSSCounterValue):
(WebCore::CSSCounterValue::create):
(WebCore::CSSCounterValue::equals const):
(WebCore::CSSCounterValue::customCSSText const):
(WebCore::CSSCounterValue::counterStyleCSSText const):
* Source/WebCore/css/CSSCounterValue.h:
(WebCore::CSSCounterValue::counterStyle const):
(WebCore::CSSCounterValue::listStyle const): Deleted.
* Source/WebCore/css/DeprecatedCSSOMCounter.h:
* Source/WebCore/css/DeprecatedCSSOMPrimitiveValue.cpp:
(WebCore::DeprecatedCSSOMPrimitiveValue::getCounterValue const):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeCounterContent):
(WebCore::CSSPropertyParserHelpers::consumeContent):
* Source/WebCore/rendering/RenderCounter.cpp:
(WebCore::RenderCounter::originalText const):
(WebCore::RenderCounter::counterStyle const):
* Source/WebCore/rendering/RenderCounter.h:
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::listMarkerText):
(WebCore::RenderListMarker::counterStyle const):
* Source/WebCore/rendering/style/CounterContent.h:
(WebCore::CounterContent::CounterContent):
(WebCore::CounterContent::listStyleType const):
(WebCore::operator==):
(WebCore::CounterContent::listStyle const): Deleted.
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueContent):

Canonical link: https://commits.webkit.org/261985@main
